schnorr_fun/src/frost/share.rs

```diff
@@ -0,0 +1,297 @@
+
+    impl fmt::Display for ShareDecodeError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            match &self {
+                ShareDecodeError::Bech32DecodeError(e) => {
+                    write!(f, "Failed to decode bech32m string: {e}")
+                }
+                ShareDecodeError::InvalidSecretShareScalar => {
+                    write!(
+                        f,
+                        "Invalid secret share scalar value, not on secp256k1 curve."
+                    )
+                }
+                ShareDecodeError::InvalidHumanReadablePrefix => {
+                    write!(f, "Expected human readable prefix `frost`",)
+                }
+                ShareDecodeError::InvalidShareIndexScalar => {
+                    write!(f, "Share index scalar was not a valid secp256k1 scalar.",)
+                }
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+        use crate::frost::SecretShare;
+        use alloc::string::ToString;
+        use core::str::FromStr;
+        use secp256kfun::{proptest::prelude::*, Scalar};
+
+        proptest! {
+            #[test]
+            fn share_backup_roundtrip(index in any::<Scalar<Public, NonZero>>(), secret in any::<Scalar<Secret, Zero>>()) {
+                let orig = SecretShare { secret, index };
+                let orig_encoded = orig.to_string();
+                let decoded = SecretShare::from_str(&orig_encoded).unwrap();
+                assert_eq!(orig, decoded)
+            }
+
+
+            #[test]
+            fn short_backup_length(secret in any::<Scalar<Secret, Zero>>(), share_index_u32 in 1u32..200) {
+                let index = Scalar::<Public, Zero>::from(share_index_u32).non_zero().unwrap().public();
+                let secret_share = SecretShare {
+                    index,
+                    secret,
+                };
+                let backup = secret_share
+                .to_string();
+
+                if share_index_u32 >= HUMAN_READABLE_THRESHOLD {
+                    assert!(backup.starts_with("frost1"));
+                } else {
+                    assert!(backup.starts_with(&format!("frost[{}]", share_index_u32)));
+                }
+
+                assert_eq!(SecretShare::from_str(&backup), Ok(secret_share))
+            }
+        }
+    }
+}
+
+#[cfg(feature = "share_backup")]
+pub use share_backup::ShareDecodeError;
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::frost;
+    use alloc::vec::Vec;
+    use secp256kfun::{
+        g,
+        proptest::{
+            prelude::*,
+            test_runner::{RngAlgorithm, TestRng},
+        },
+        G,
+    };
+    proptest! {
+        #[test]
+        fn recover_secret(parties in 1usize..10, threshold in 1usize..5) {
+            use rand::seq::SliceRandom;
+            let frost = frost::new_with_deterministic_nonces::<sha2::Sha256>();
+            let parties = parties.max(threshold);
```

Comment:
could use `(parties, threshold) in (1usize..=10).prop_flat_map(|n| (Just(n), 1usize..=n))` but it's also kind of confusing